src/NuGet.Core/NuGet.Packaging/PackageExtraction/ZipArchiveExtensions.cs

```diff
@@ -65,6 +66,19 @@ public static string SaveAsFile(this ZipArchiveEntry entry, string fileFullPath,
             return fileFullPath;
         }
 
+        private static readonly int UpdateFileTimeFromEntryMaxRetries = GetUpdateFileTimeFromEntryMaxRetries();
+
+        private static int GetUpdateFileTimeFromEntryMaxRetries()
+        {
+            string value = Environment.GetEnvironmentVariable("NUGET_UpdateFileTime_MaxRetries");
```

Comment:
Thank you for addressing the above comment. I am sorry if this comment contradicts my above comment. AFAIK, it makes sense to add a unique environment variable to turn off/on a feature. I started questioning myself if the environment variable is really required in this case because it is hard to predict the correct value as it depends mostly on the timing.

It looks like `dotnet installer` repo has a same problem but interestingly comment says `Memory-mapped write is not updating last write time`. Maybe we are facing this issue because the product uses memory mapped files. The team ended up adding retry logic to set the file timestamp as well.

https://github.com/dotnet/runtime/blob/d0631be3a1b9e5a8adf16f4885778179a8ed365b/src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs#L203-L205

The retry logic has a comment that says the following and the [code](https://github.com/dotnet/runtime/blob/d0631be3a1b9e5a8adf16f4885778179a8ed365b/src/installer/managed/Microsoft.NET.HostModel/AppHost/RetryUtil.cs#L26-L39) simply retires a fixed number of times with 100 milliseconds delay between retries.

``` xml
     /// An Antivirus scanner may intercept in-between and lock the file,
    /// causing the operations to fail with IO-Error. `
    /// So, the operations are retried a few times on failures such as
    /// - IOException
    /// - Failure with Win32 errors indicating file-lock
```

What do you think about fixing this issue without an environment variable? It is very much possible that I may have missed some scenarios while providing this feedback.

Reply:
>  I started questioning myself if the environment variable is really required in this case because it is hard to predict the correct value as it depends mostly on the timing.

That's precisely why I added the environment variable. I don't know if retrying over 10ms or 100ms or 1s is ideal, so I made a guess, and if it's not a good guess, I provide customers a way to adjust it. If we get feedback in the future, we can adjust the default.

> Maybe we are facing this issue because the product uses memory mapped files.

How would mmapio not updating the last write time cause an `IOException` when we set the file's last write time to the timstamp in the zip, after we've closed the file?

> What do you think about fixing this issue without an environment variable?

The .NET installer is a small and very specific program, whose only responsibility is to install a set of files. NuGet restores multiple projects, down parallel downloads, extracts multiple packages, does asset selection, and maybe more. I think putting extraction in a busy loop retrying to write the file until it succeeds, prevents NuGet from doing other work, hence why I implemented an exponential backoff. If it can't progress quickly, then wait longer and longer between retries, so that, when there's a lot of other work to do, we don't burn one CPU core.

So, I'm not a fan of taking the same approach as the .NET installer, because it feels to me like a quick-and-easy mitigation where performance doesn't matter. I don't consider it evidence of a best practice or recommended pattern. For NuGet, restore performance does matter, so I think exponential backoff makes sense, and the only question I have is for how long to keep retrying before giving up.

Reply:
I agree that increasing wait time exponentially between retries is ideal considering restoring performance is critical. I am curious about how customers can predict value for the environment variable because the value may work once but not later due to the timing. Given that max retries is updated to `9` in the latest commit, I guess it should be sufficient to get a file handle for updating the last time stamp without the need to adjust the retries count via environment variable.

> How would mmapio not updating the last write time cause an IOException when we set the file's last write time to the timestamp in the zip, after we've closed the file?

I am a beginner in mmap io files. Please correct if my understanding is correct. `CopyToFile` method (see below code map image) writes to mmap stream which is invoked during package extraction. If my understanding is correct, we try to update the timestamp after copying to mmap stream. I don't know but guessing if there is any chance where mmap logic is updating the file on disk and NuGet tries to update the timestamp at the same time resulting in `IO Exception`.

![image](https://user-images.githubusercontent.com/52756182/148126281-84552835-523a-4735-ae88-4d026589b5d8.png)

For e.g. one code path which extracts file (sometimes it may copy to mmap stream) and immediately tries to update the timestamp.

https://github.com/NuGet/NuGet.Client/blob/e050b00c3683d3ecd18a8885ef464ec32b1e86b7/src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs#L322-L325

Looking at the above code map, there is another code path that tries to call `File.SetLastWriteTimeUtc` method. I am unsure at which stage during restore/installation this code path will be executed. Should we add retry logic here as well?

https://github.com/NuGet/NuGet.Client/blob/e050b00c3683d3ecd18a8885ef464ec32b1e86b7/src/NuGet.Core/NuGet.Packaging/PackageFolderReader.cs#L213-L220

Reply:
I see from your edit that you noticed that we only update the last write time after we close the file, and closing the mmap file will flush to disk, hence we can be confident that our own process does not have the file open at the time we try to set the last write time.

Is this comment thread still relevant? I don't understand what actions I might need to take.

Reply:
@zivkan - The only action item I can think of that is still relevant is the following one.

Looking at the above code map, there is another code path that tries to call `File.SetLastWriteTimeUtc` method. I am unsure at which stage during restore/installation this code path will be executed. Should we add retry logic here as well?

https://github.com/NuGet/NuGet.Client/blob/e050b00c3683d3ecd18a8885ef464ec32b1e86b7/src/NuGet.Core/NuGet.Packaging/PackageFolderReader.cs#L213-L220

Reply:
thanks @kartheekp-ms. I pushed an update which makes this use the same retry logic.